src/hazardous/aead/xchacha20poly1305_stream.rs

```diff
@@ -0,0 +1,1361 @@
+		const_assert!(SECRETSTREAM_XCHACHA20POLY1305_INONCEBYTES == core::mem::size_of::<INonce>());
+		const_assert!(
+			SECRETSTREAM_XCHACHA20POLY1305_INONCEBYTES
+				+ SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES
+				== IETF_CHACHA_NONCESIZE
+		);
+
+		let mut inonce = [0u8; SECRETSTREAM_XCHACHA20POLY1305_INONCEBYTES];
+		inonce.copy_from_slice(&nonce.as_ref()[HCHACHA_NONCESIZE..]);
+
+		Self {
+			key: subkey_and_nonce(&secret_key, &nonce).0,
+			counter: 1,
+			inonce,
+		}
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// Derives a new secret key used for encryption and decryption.
+	pub fn rekey(&mut self) -> Result<(), UnknownCryptoError> {
+		let mut new_key_and_inonce =
+			[0u8; CHACHA_KEYSIZE + SECRETSTREAM_XCHACHA20POLY1305_INONCEBYTES];
+		new_key_and_inonce[..CHACHA_KEYSIZE].copy_from_slice(self.key.unprotected_as_bytes());
+		new_key_and_inonce[CHACHA_KEYSIZE..].copy_from_slice(&self.inonce);
+
+		chacha20_xor_stream(&self.key, &self.get_nonce(), 0, &mut new_key_and_inonce)?;
+
+		self.key = SecretKey::from_slice(&new_key_and_inonce[..CHACHA_KEYSIZE]).unwrap();
+		self.inonce
+			.copy_from_slice(&new_key_and_inonce[CHACHA_KEYSIZE..]);
+		self.counter = 1;
+
+		Ok(())
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// Encrypt and authenticate a single message and tag.
+	pub fn seal_chunk(
+		&mut self,
+		plaintext: &[u8],
+		ad: Option<&[u8]>,
+		dst_out: &mut [u8],
+		tag: Tag,
+	) -> Result<(), UnknownCryptoError> {
+		// The largest addition done on length
+		// of plaintext. Other non-checked additions
+		// should not be able to overflow if this
+		// does not panic.
+		assert!((CHACHA_BLOCKSIZE as u64)
+			.checked_add(plaintext.len() as u64)
+			.is_some());
+
+		let msglen = plaintext.len();
+		if dst_out.len() < SECRETSTREAM_XCHACHA20POLY1305_ABYTES + msglen {
+			return Err(UnknownCryptoError);
+		}
+
+		let mut block = [0u8; CHACHA_BLOCKSIZE];
+		let ad = match ad {
+			Some(v) => v,
+			None => &[0u8; 0],
+		};
+		let cipherpos = core::mem::size_of::<Tag>();
+		let macpos = cipherpos + msglen;
+
+		block[0] = tag.bits();
+		chacha20_xor_stream(&self.key, &self.get_nonce(), 1, &mut block)?;
+		dst_out[0] = block[0];
+
+		if msglen != 0 {
+			chacha20_enc(
+				&self.key,
+				&self.get_nonce(),
+				2,
+				plaintext,
+				&mut dst_out[cipherpos..],
+			)?;
+		}
+
+		let mac = self.generate_auth_tag(dst_out, ad, msglen, &block, cipherpos)?;
+
+		debug_assert!(dst_out.len() >= macpos + mac.get_length());
+		dst_out[macpos..(macpos + POLY1305_OUTSIZE)].copy_from_slice(mac.unprotected_as_bytes());
+
+		xor_slices_8(
+			self.inonce.as_mut(),
+			&dst_out[macpos..macpos + SECRETSTREAM_XCHACHA20POLY1305_INONCEBYTES],
+		);
+		self.counter = self.counter.wrapping_add(1);
+
+		if bool::from(tag.bits().ct_eq(&Tag::REKEY.bits()) | self.counter.ct_eq(&0u32)) {
```

Comment:
This check will never trigger a rekey on `Tag::FINISH` due to comparison on its bits instead of using bitwise-AND. The new tests I've added should also detect this.

This has to be changed to something in the lines of:
```
bool::from(!(tag.bits() & Tag::REKEY.bits()).ct_eq(&0u8) | self.counter.ct_eq(&0u32))
```

Similarly in `open_chunk`.